src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java

```diff
@@ -444,25 +448,27 @@ protected Collection<Object> _deserializeWithObjectId(JsonParser p, Deserializat
         // [databind#631]: Assign current value, to be accessible by custom serializers
         p.assignCurrentValue(result);
 
-        final JsonDeserializer<Object> valueDes = _valueDeserializer;
-        final TypeDeserializer typeDeser = _valueTypeDeserializer;
         CollectionReferringAccumulator referringAccumulator =
                 new CollectionReferringAccumulator(_containerType.getContentType().getRawClass(), result);
 
         JsonToken t;
         while ((t = p.nextToken()) != JsonToken.END_ARRAY) {
             try {
                 Object value;
+
                 if (t == JsonToken.VALUE_NULL) {
                     if (_skipNullValues) {
                         continue;
                     }
-                    value = _nullProvider.getNullValue(ctxt);
-                } else if (typeDeser == null) {
-                    value = valueDes.deserialize(p, ctxt);
+                    value = null;
                 } else {
-                    value = valueDes.deserializeWithType(p, ctxt, typeDeser);
+                    value = _deserializeRawContentValue(p, ctxt);
+                }
+
+                if (value == null) {
+                    value = _nullProvider.getNullValue(ctxt);
```

Comment:
Is it appropriate to call `_nullProvider` on a deserialized result using `_valueDeserializer`?

Reply:
I don't think so; it should be applied to input `null`s. Although TBH, value deserializers are also not expected to be handed `null` tokens ever, hence should rarely return `null`s too.

So no should not call on deserialized values.

Reply:
Actually, right. Based on fact `null` comes via coercion, I guess this must be done. No real way around it: null check/handling needed in separate cases.